### PR TITLE
Added Transaction Indexing

### PR DESCRIPTION
**Pull request by sandipchandra**
## Link to JIRA ticket

Add a link to the JIRA ticket that the changes in this PR describe.

### *What is the problem we're trying to solve*

When ever there is a transaction is generated we also need to index the same in Elastic Search.

### *What changes have we introduced*

When ever there is a transaction is generated we also need to index the same in Elastic Search. We have added use case in Housing Search Listener to index Transaction in ES so that Housing Search API Transaction search endpoint will work for latest Transaction.

Include any links to commits that introduce significant additions of code if they help explain the process of coming to the solution. e.g Addition of test database setup, addition of first test and production class, removal of buggy code, etc.

#### _Checklist_

- [checked ] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [checked  ] Added tests to cover all new production code
- [ checked ] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [ checked ] Checked all code for possible refactoring
- [ checked ] Code pipeline builds correctly

### *Follow up actions after merging PR*

Are there any next steps that need to addressed after merging this PR? Add them here.

### CODE REVIEW

**Review by EdwardAnoghena — APPROVED** [review]

**Review by HHjolany — APPROVED** [review]

File: HousingSearchListener/V1/UseCase/Interfaces/IIndexTransactionUseCase.cs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+﻿namespace HousingSearchListener.V1.UseCase.Interfaces
+{
+    public interface IIndexTransactionUseCase : IMessageProcessing
+    {
```
Comment (humulla) [inline HousingSearchListener/V1/UseCase/Interfaces/IIndexTransactionUseCase.cs:4]: This interface should not be empty it should have this method in it:
`Task ProcessMessageAsync(EntityEventSns message)`

File: HousingSearchListener/serverless.yml [hunk 2]
```diff
@@ -26,6 +26,8 @@ functions:
       TenureApiToken: ${ssm:/housing-tl/${self:provider.stage}/tenure-api-token}
       AccountApiUrl: ${ssm:/housing-finance/${self:provider.stage}/account-api-url}
       AccountApiToken: ${ssm:/housing-tl/${self:provider.stage}/account-api-token}
+      FinancialTransactionApiUrl: ${ssm:/housing-finance/${self:provider.stage}/financial-transaction-api-url}
+      FinancialTransactionApiToken: ${ssm:/housing-tl/${self:provider.stage}/financial-transaction-api-token}
```
Comment (humulla) [inline HousingSearchListener/serverless.yml:30]: the parameter name should start with `housing-finance `

File: HousingSearchListener/V1/UseCase/Interfaces/IIndexTransactionUseCase.cs (hunk 1 above)
Comment (sandipchandra) [reply HousingSearchListener/V1/UseCase/Interfaces/IIndexTransactionUseCase.cs:4]: This method is already there in base interface and if you compare with te other use cases interfaces it also does not have this method defined again.

File: HousingSearchListener.Tests/V1/Factories/ESEntityFactoryTests.cs [hunk 3]
```diff
@@ -112,5 +114,38 @@ public void CreateAssetQueryableTenureTest()
             result.StartOfTenureDate.Should().Be(domainTenure.StartOfTenureDate);
             result.Type.Should().Be(domainTenure.TenureType.Description);
         }
+        [Fact]
+        public void CreateQueryableTransactionTest()
+        {
+            var domainTransaction = _fixture.Create<TransactionResponseObject>();
+
+            var result = _sut.CreateQueryableTransaction(domainTransaction);
+            result.Address.Should().Be(domainTransaction.Address);
```
[the remark this answers is not in the document]
Comment (sandipchandra) [reply HousingSearchListener.Tests/V1/Factories/ESEntityFactoryTests.cs:123]: result is of type "QueryableTransaction" which is from shared library gateway model and domainTransaction in of type TransactionResponseObject which is defined locally inside project.

[the remark this answers is not in the document]
Comment (sandipchandra) [reply HousingSearchListener.Tests/V1/Factories/ESEntityFactoryTests.cs:123]: I tried with this approach but due to diff object type the unit test case was failing.

File: HousingSearchListener/V1/Domain/Transaction/Person.cs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace HousingSearchListener.V1.Domain.Transaction
+{
+    public class Person
```
[the remark this answers is not in the document]
Comment (sandipchandra) [reply HousingSearchListener/V1/Domain/Transaction/Person.cs:5]: Housing.Shared repo is already having a domain model for Transaction but as requested by Mirela we had to rename the Person object to Sender there for search result.
But in Financial Transaction Api response its mentioned as Person object. Hence we cannot re use the shared model in this scenario.

File: HousingSearchListener/V1/Domain/Transaction/TransactionResponseObject.cs [hunk 5]
```diff
@@ -0,0 +1,81 @@
+﻿using System;
+using Hackney.Shared.HousingSearch.Domain.Transactions;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Converters;
+
+namespace HousingSearchListener.V1.Domain.Transaction
+{
+    public class TransactionResponseObject
```
[the remark this answers is not in the document]
Comment (sandipchandra) [reply HousingSearchListener/V1/Domain/Transaction/TransactionResponseObject.cs:8]: Housing.Shared repo is already having a domain model for Transaction but as requested by Mirela we had to rename the Person object to Sender there for search result.
But in Financial Transaction Api response its mentioned as Person object. Hence we cannot re use the shared model in this scenario

File: HousingSearchListener/serverless.yml (hunk 2 above)
Comment (humulla) [reply HousingSearchListener/serverless.yml:30]: Maybe @LBHSPreston created it?

Comment (sandipchandra) [reply HousingSearchListener/serverless.yml:30]: I think so

Comment (humulla) [reply HousingSearchListener/serverless.yml:30]: @LBHSPreston when you get a chance can you please confirm with the financial transactions api token parameter name should be, I think it should be `housing-finance/….` but want to make sure

Comment (sandipchandra) [reply HousingSearchListener/serverless.yml]: @humulla  I have modified the SSM parameter name, please have a look and approve the PR.

**Review by humulla — APPROVED** [review]

**Pull request merged**